File: lib/data/models/poker.dart
import 'package:flutter/material.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:landlords_3/core/card/card_type.dart';
part 'poker.g.dart';

@JsonSerializable()
class Poker {
  @JsonKey(name: 'suit', toJson: _suitToJson, fromJson: _suitFromJson)
  final Suit suit;
  @JsonKey(name: 'value', toJson: _valueToJson, fromJson: _valueFromJson)
  final CardValue value;

  Poker({required this.suit, required this.value});

  factory Poker.fromJson(Map<String, dynamic> json) => _$PokerFromJson(json);
  Map<String, dynamic> toJson() => _$PokerToJson(this);

  // 枚举转换方法
  static String _suitToJson(Suit suit) => suit.name;
  static Suit _suitFromJson(String json) =>
      Suit.values.firstWhere((e) => e.name == json, orElse: () => Suit.joker);

  static String _valueToJson(CardValue value) => value.name;
  static CardValue _valueFromJson(String json) => CardValue.values.firstWhere(
    (e) => e.name == json,
    orElse: () => CardValue.jokerBig,
  );

  // 用于显示的文字
  String get displayValue {
    switch (value) {
      case CardValue.ace:
        return 'A';
      case CardValue.two:
        return '2';
      case CardValue.three:
        return '3';
      case CardValue.four:
        return '4';
      case CardValue.five:
        return '5';
      case CardValue.six:
        return '6';
      case CardValue.seven:
        return '7';
      case CardValue.eight:
        return '8';
      case CardValue.nine:
        return '9';
      case CardValue.ten:
        return 'X';
      case CardValue.jack:
        return 'J';
      case CardValue.queen:
        return 'Q';
      case CardValue.king:
        return 'K';
      case CardValue.jokerSmall:
        return 'joker';
      case CardValue.jokerBig:
        return 'Joker';
      default:
        return '';
    }
  }

  // 获取颜色
  Color get color {
    if (suit == Suit.hearts || suit == Suit.diamonds) {
      return Colors.red;
    } else if (suit == Suit.joker) {
      return Colors.red; // 可以根据大小王设置不同的颜色
    } else {
      return Colors.black;
    }
  }

  // 获取花色符号
  String get suitSymbol {
    switch (suit) {
      case Suit.hearts:
        return '♥';
      case Suit.diamonds:
        return '♦';
      case Suit.clubs:
        return '♣';
      case Suit.spades:
        return '♠';
      default:
        return ''; // 大小王没有花色符号
    }
  }
}

File: server_3/src/models/Poker.ts
import { CardValue, Suit } from "../card/CardType";
import { CardUtils } from "../card/CardUtils";


export class Poker {
    constructor(
        public readonly suit: Suit,
        public readonly value: CardValue
    ) { }

    // 权重值用于比较大小（后端核心逻辑需要）
    get weight(): number {
        return CardUtils.getCardWeight(this.value);
    }

    // 序列化方法（用于网络传输）
    toJSON() {
        return {
            suit: Suit[this.suit],
            value: CardValue[this.value]
        };
    }
    static fromJSON(json: { suit: string; value: string }): Poker {
        // 转换花色字符串为枚举值
        const suitMapping: { [key: string]: Suit } = {
            'hearts': Suit.hearts,
            'diamonds': Suit.diamonds,
            'clubs': Suit.clubs,
            'spades': Suit.spades,
            'joker': Suit.joker
        };

        // 转换牌值字符串为枚举值
        const valueMapping: { [key: string]: CardValue } = {
            'three': CardValue.three,
            'four': CardValue.four,
            'five': CardValue.five,
            'six': CardValue.six,
            'seven': CardValue.seven,
            'eight': CardValue.eight,
            'nine': CardValue.nine,
            'ten': CardValue.ten,
            'jack': CardValue.jack,
            'queen': CardValue.queen,
            'king': CardValue.king,
            'ace': CardValue.ace,
            'two': CardValue.two,
            'jokerSmall': CardValue.jokerSmall,
            'jokerBig': CardValue.jokerBig
        }
        return new Poker(suitMapping[json.suit], valueMapping[json.value])
    }
    // 比较两张牌的大小
    compareTo(other: Poker): number {
        return this.weight - other.weight;
    }
}